### DISCUSSION TITLE
Exposing FileLocationSettings

### DISCUSSION THREAD

**Discussion by RhodeIfado** [BioMotionLab/TUX#76] [Ideas]
Hey Adam,

I've got a question about a feature request I'd like to work on, but I wanted to ask if you could share your thoughts. Maybe this idea has been considered but scrapped for reasons.

My goal is to make the location for all the data more flexibel, rather than having to save into the Documents\bmltux_data folder. When I want to change the Location I have to edit the package because FileLocationSettings.cs isn't exposed in the editor, unlike other Settings (ControlSettings, ColumnNameSettings ... ). I've been using a modified local package of this project for a while now because I've made other changes aswell, but this change might be useful for everyone.

As far as I understand, the problem stems from FileLocationSettings beeing a static class and static classes / variables are unserializable. Additonaly its supposed to be read-only, which complicates things alot. 
I've tinkered around with a proxy serialization class, for serialization and protected inheritance access, but thats not going in the right direction. Would it be more sensible to try to modify the FileLocationSettings to behave like an Object thats passed to ExperimentSettings and I/ExperimentDesignFile2 ?

Regards
Jakob

**Comment by AdamBebko** [comment]
Hi Jakob, I think your suggestion would be a welcome addition, to allow users to override the default settings using a scriptable object that is passed into the experiment config file. 

If you decide to work on this, just keep in mind a couple things

- Current behaviour should stay consistent if a user doesn't change anything (no breaking changes)
- The user should not have to do anything to use default settings, and do something to override them.

Note, on many devices, you aren't allowed to save files save anywhere, and are are forced to use the unity-provided folders. I know windows and Mac allow you to save anywhere, but mobile does not (including meta quest if running in standalone), so just something to keep in mind.

Let me know if you need a hand with anything.

**Comment by RhodeIfado** [comment]
I've got a question regarding a the PackageInstallDetector which accesses the FileLocationSettings.
During the setup of the filepath its using the BaseTuxDocumentsFolderPath, but I dont quite understand why.

https://github.com/BioMotionLab/TUX/blob/e1514d88490c6e36167758a3168434985e2422c2/BML_TUX_Project/Packages/bmlTUX/Editor/PackageInstallDetector.cs#L57-L58

Passing the BaseTuxDocumentsFolderPath into Path.Combine will never have an effect, because a following argument contains a rooted path, in this case the Application.dataPath ([ref](https://learn.microsoft.com/en-gb/dotnet/api/system.io.path.combine?view=net-8.0)).

Testing this, there have been no occasions of the installData.json ever being created in the BaseTuxDocumentsFolderPath and I couldn't think of a case when it should be.

I bring up this specific situation, because PackageInstallDetector is an Editor Script which executes immediatly upon Asset Import. I think its not as straight forward as I'd hoped to link the Scriptable FileLocationSettings into this Script without further changes. Removing the reference would make things alot easier and wouldnt require any changes to this Scripts Behaviour, because the reference doesnt have any effect anyways.

**Reply by AdamBebko** [reply]
I agree that this looks like a bug. I think it’s safe to always save the install data to Application.datapath. And that can be independent of FileLocationSettings.

**Reply by AdamBebko** [reply]
Not that fixing this will cause all users to receive the pop up once. But I think that’s OK for more long-term correctness.
